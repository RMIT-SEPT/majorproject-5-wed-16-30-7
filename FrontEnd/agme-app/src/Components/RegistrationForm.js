import React from 'react';
import './LoginForm.scss';

class RegistrationForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            name: '',
            address: '',
            phone: '',
            username: '',
            password: '',
            confirmPass: '',

            errors: {
                name: '',
                address: '',
                phone: '',
                username: '',
                password: '',
                confirmPass: '',
            }
        };

        this.handleChange = this.handleChange.bind(this);
        this.getCustomerRegistration = this.getCustomerRegistration.bind(this);
    }


    getCustomerRegistration = (e) => {
        e.preventDefault();

        //Check if form contains any errors
        const validateForm = (errors) => {
            let valid = true;
            Object.values(errors).forEach(
                (val) => val.length > 0 && (valid = false)
            );
            return valid;
        }

        //Submit form if form is valid
        if (validateForm(this.state.errors)) {
            console.info('Valid Form');
            let customer = {
                name: this.state.name,
                address: this.state.address,
                phone: this.state.phone,
                username: this.state.username,
                password: this.state.password
            };
            console.log('customerLogin =>' + JSON.stringify(customer));
        } else {
            console.error('Invalid Form')
        }

    }


    handleChange = (event) => {
        event.preventDefault();
        const { name, value } = event.target;
        let errors = this.state.errors;

        switch (name) {
            case 'name':
                errors.name =
                    value.length > 25
                        ? 'Name too long!'
                        : '';
                break;
            case 'username':
                errors.username =
                    value.length < 5
                        ? 'Username too short!'
                        : '';
                break;
            case 'password':
                errors.password =
                    value.length < 8
                        ? 'Password must be atleast 8 characters long!'
                        : '';
                break;
            case 'confirmPass':
                errors.confirmPass =
                    value !== this.state.password
                        ? 'Password does not match!'
                        : '';
                break;
            default:
                break;
        }

        this.setState({ errors, [name]: value }, () => {
            // console.log(errors)
        })
    }

    render() {
        const { errors } = this.state;
        return (
            <form id="register" className="input-group">
                <input type="text" className="input-field" name='name' placeholder="Full Name" value={this.state.name} onChange={this.handleChange} required />
                {errors.name.length > 0 &&
                    <span className='error'>{errors.name}</span>}
                <input type="text" className="input-field" name='address' placeholder="Address" value={this.state.address} onChange={this.handleChange} required />
                <input type="text" className="input-field" name='phone' placeholder="Phone Number" value={this.state.phone} onChange={this.handleChange} required />
                <input type="text" className="input-field" name='username' placeholder="User Name" value={this.state.username} onChange={this.handleChange} required />
                {errors.username.length > 0 &&
                    <span className='error'>{errors.username}</span>}
                <input type="password" className="input-field" name='password' placeholder="Password" value={this.state.password} onChange={this.handleChange} required />
                {errors.password.length > 0 &&
                    <span className='error'>{errors.password}</span>}
                <input type="password" className="input-field" name='confirmPass' placeholder="Confirm Password" value={this.state.confirmPass} onChange={this.handleChange} required />
                {errors.confirmPass.length > 0 &&
                    <span className='error'>{errors.confirmPass}</span>}
                <button type="submit" className="submit-btn" onClick={this.getCustomerRegistration}>Register</button>
            </form>
        );
    }
}

export default RegistrationForm;