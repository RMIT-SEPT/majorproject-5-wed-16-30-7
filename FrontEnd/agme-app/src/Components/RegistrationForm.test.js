import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';


import RegistrationForm from './RegistrationForm.js';


let container = null;
beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

describe("Render form", () => {
    it("Should render registration form", () => {
        act(() => {
            render(<RegistrationForm />, container);
        });
        expect(container).toMatchSnapshot();

    });
})

describe("Input Values", () => {
    const onChange = jest.fn();
    const testNames = ["name", "address", "phone", "username", "password", "confirmPass"];

    for (let i = 0; i < testNames.length; ++i) {
        it("Updates " + testNames[i] + " field on change", () => {
            act(() => {
                render(<RegistrationForm onChange={onChange} />, container);
            });
            var input = document.querySelector("[data-testid=" + CSS.escape(testNames[i].toString()) + "]");
            fireEvent.change(input, { target: { value: testNames[i] } });
            expect(input.value).toBe(testNames[i]);
        });
    }

});

// describe("Check handleChange function and errors states", () => {
//     const handleChange = jest.spyOn(RegistrationForm, 'handleChange');
//     const evnetTriggered = RegistrationForm.handleChange();

//     expect(handleChange).toHaveBeenCalled();
//     expect(evnetTriggered).toBe(true);
// })
