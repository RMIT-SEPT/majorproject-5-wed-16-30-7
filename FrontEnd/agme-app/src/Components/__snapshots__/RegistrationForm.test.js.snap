// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render form Should render registration form 1`] = `
<div>
  <form
    class="input-group"
    id="register"
  >
    <input
      class="input-field"
      data-testid="name"
      name="name"
      placeholder="Full Name"
      required=""
      type="text"
      value=""
    />
    <input
      class="input-field"
      data-testid="address"
      name="address"
      placeholder="Address"
      required=""
      type="text"
      value=""
    />
    <input
      class="input-field"
      data-testid="phone"
      name="phone"
      placeholder="Phone Number"
      required=""
      type="text"
      value=""
    />
    <input
      class="input-field"
      data-testid="username"
      name="username"
      placeholder="User Name"
      required=""
      type="text"
      value=""
    />
    <input
      class="input-field"
      data-testid="password"
      name="password"
      placeholder="Password"
      required=""
      type="password"
      value=""
    />
    <input
      class="input-field"
      data-testid="confirmPass"
      name="confirmPass"
      placeholder="Confirm Password"
      required=""
      type="password"
      value=""
    />
    <button
      class="submit-btn"
      type="submit"
    >
      Register
    </button>
  </form>
</div>
`;
